src/XR/features/WebXRNearInteraction.ts

```diff
@@ -0,0 +1,590 @@
+     * @param _options read-only options to be used in this module
+     */
+    constructor(_xrSessionManager: WebXRSessionManager, private readonly _options: IWebXRNearInteractionOptions) {
+        super(_xrSessionManager);
+        this._scene = this._xrSessionManager.scene;
+    }
+
+    /**
+     * attach this feature
+     * Will usually be called by the features manager
+     *
+     * @returns true if successful.
+     */
+    public attach(): boolean {
+        if (!super.attach()) {
+            return false;
+        }
+
+        this._options.xrInput.controllers.forEach(this._attachController);
+        this._addNewAttachObserver(this._options.xrInput.onControllerAddedObservable, this._attachController);
+        this._addNewAttachObserver(this._options.xrInput.onControllerRemovedObservable, (controller) => {
+            // REMOVE the controller
+            this._detachController(controller.uniqueId);
+        });
+
+        this._scene.constantlyUpdateMeshUnderPointer = true;
+        return true;
+    }
+
+    /**
+     * detach this feature.
+     * Will usually be called by the features manager
+     *
+     * @returns true if successful.
+     */
+    public detach(): boolean {
+        if (!super.detach()) {
+            return false;
+        }
+
+        Object.keys(this._controllers).forEach((controllerId) => {
+            this._detachController(controllerId);
+        });
+
+        return true;
+    }
+
+    /**
+     * Will get the mesh under a specific pointer.
+     * `scene.meshUnderPointer` will only return one mesh - either left or right.
+     * @param controllerId the controllerId to check
+     * @returns The mesh under pointer or null if no mesh is under the pointer
+     */
+    public getMeshUnderPointer(controllerId: string): Nullable<AbstractMesh> {
+        if (this._controllers[controllerId]) {
+            return this._controllers[controllerId].meshUnderPointer;
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Get the xr controller that correlates to the pointer id in the pointer event
+     *
+     * @param id the pointer id to search for
+     * @returns the controller that correlates to this id or null if not found
+     */
+    public getXRControllerByPointerId(id: number): Nullable<WebXRInputSource> {
+        const keys = Object.keys(this._controllers);
+
+        for (let i = 0; i < keys.length; ++i) {
+            if (this._controllers[keys[i]].id === id) {
+                return this._controllers[keys[i]].xrController || null;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * This function sets webXRControllerPointer Selection feature that will be disabled when
+     * the hover range is reached for a mesh and will be reattached when not in hover range.
+     * This is used to remove the selection rays when moving.
+     * @param farInteractionFeature the feature to disable when finger is in hover range for a mesh

```

Comment:
>e feature to disable when finger is in hover range for a mesh [](http://example.com/codeflow?start=38&length=61)

This is fine, though I'm not the biggest fan on how vaguely it refers to the pointerSelection feature here (since I doubt this would ever be used with a different feature).

Reply:
I understand that it is possible to use other abstract feature in place of this. But the reason i had it as an abstract feature is because of detach problems that occurred when I confine it just to be a pointer selection class. Let me add in a check inside the class to make sure it is actually controller pointer selection feature that is passed in.

Reply:
If it's expected that other classes might be used instead of the pointer feature, then it's fine to leave it as-is

Reply:
I find it the best as it is. Resolving this.